### PR TITLE
Move OpenGL App code into renderer

### PR DESCRIPTION
**Pull request by fxtressia**
So, you would like to use Inox2d! However, you then have to copy all the Glutin code from the examples. With this PR, there is a new `inox2d::renderer::App` trait that's implemented by the OpenGL renderer backend as `inox2d::renderer::opengl::app::App`. The code is more or less copied from the `render_inp` example. 

The trait is also intended to be implemented by the future `wgpu` backend! Since obviously, `wgpu` doesn't use Glutin.

# TO-DO
- [x] Remove the App code from render_inx 
- [x] Store the renderer in the App and run the renderer events there?

Signed-off-by: Fortressia <fortressnordlys@outlook.com>

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,6 +7,10 @@ edition = "2021"
 
 [dependencies]
 glam = { version = "0.22.0", features = ["serde"] }
+glutin = { version = "0.30.3", optional = true }
+raw-window-handle = "0.5.0"
+glutin-winit = { version = "0.2.1", optional = true }
+winit = { version = "0.27.5", optional = true }
```
Comment (Speykious) [inline Cargo.toml]: I see that `raw-window-handle` is not optional, is that intended? Logically it should be kept for the rendering feature.

File: Cargo.toml [hunk 2]
```diff
@@ -18,19 +22,19 @@ serde_indextree = "0.2.0"
...
-tracing-subscriber = "0.3.16"
-winit = "0.27.5"
+
 
 [features]
-default = ["opengl"]
+default = ["gl-winit", "wgpu"]
 opengl = ["dep:glow"]
+gl-winit = ["opengl", "glutin", "glutin-winit", "winit"]
+wgpu = ["dep:wgpu"]
```
Comment (Speykious) [inline Cargo.toml]: Hmmmm... I wonder if there's a better way to organize these features. This feels a bit messy.

Also I don't think `gl-winit` and `wgpu` should be both toggled at the same time. I think ideally they should be exclusive, if that makes sense.

File: src/renderers/opengl/app.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        //
+        // In particular, since we are doing offscreen rendering we have no raw window
+        // handle to provide.
+        let context_attributes = ContextAttributesBuilder::new().build(None);
+
+        // Since glutin by default tries to create OpenGL core context, which may not be
+        // present we should try gles.
+        let fallback_context_attributes = ContextAttributesBuilder::new()
+            .with_context_api(ContextApi::Gles(None))
+            .build(None);
```
Comment (Speykious) [inline src/renderers/opengl/app.rs]: Interesting :thinking:

File: src/renderers/opengl/app.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                    glow::DEBUG_SEVERITY_NOTIFICATION => debug!(target: "opengl", "{ty}{msg}"),
+                    glow::DEBUG_SEVERITY_LOW => info!(target: "opengl", "{ty}{msg}"),
+                    glow::DEBUG_SEVERITY_MEDIUM => warn!(target: "opengl", "{ty}{msg}"),
+                    glow::DEBUG_SEVERITY_HIGH => error!(target: "opengl", "{ty}{msg}"),
+                    sevr => unreachable!("unknown debug severity {sevr}"),
+                };
+            });
+
+            gl.enable(glow::DEBUG_OUTPUT);
+        }
```
Comment (Speykious) [inline src/renderers/opengl/app.rs:159]: I wonder if we should only run this block of code in debug mode.

File: src/renderers/opengl/app.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+        unsafe { gl.viewport(0, 0, 2048, 2048) };
```
Comment (Speykious) [inline src/renderers/opengl/app.rs:161]: Just a note to myself that I have to do something about that viewport. Inochi2D has a `Camera` class to handle puppet position and zoom, but I have none of that currently, I'll have to see what I can do.

**Review by Speykious — CHANGES_REQUESTED** [review]
Ayyy LGTM, although I think features should be segmented better.

File: Cargo.toml (hunk 1 above)
Comment (fxtressia) [reply Cargo.toml]: It isn't intended, sorry, but I wasn't sure under what feature should it be kept under.

Comment (fxtressia) [reply Cargo.toml]: Maybe we should have a `rendering` feature that will be enabled by `wgpu` and `opengl`?

Comment (Speykious) [reply Cargo.toml]: Probably the best option, yeah. `render` for short!
`render` would enable the `winit` and `raw-window-handle` dependencies, and `opengl` would enable the `glutin` and `glutin-winit` dependencies.

Comment (Speykious) [reply Cargo.toml]: And also `opengl` and `wgpu` would both require the `render` feature.

File: Cargo.toml (hunk 2 above)
Comment (fianathedevgirl) [reply Cargo.toml:37]: I enabled all renderers for easier development. Please remind me to remove them before merging into main.

Additionally, these lines of changes were included in this PR by mistake. It should have been in #6.

File: src/renderers/opengl/app.rs (hunk 3 above)
Comment (fxtressia) [reply src/renderers/opengl/app.rs]: Yup!

File: Cargo.toml (hunk 2 above)
Comment (Speykious) [reply Cargo.toml]: _woops :v_

File: src/renderers/opengl/app.rs (hunk 4 above)
Comment (fxtressia) [reply src/renderers/opengl/app.rs:159]: I think it's much more better to run them when the `INOX_VERBOSE` environment variable is set to `1`.

File: src/renderers/opengl/app.rs (hunk 5 above)
Comment (fxtressia) [reply src/renderers/opengl/app.rs:161]: Yup! And also, I have been wondering. We need a way to render multiple puppets within one camera. 

Especially considering my use case would be for an anime RPG game with puppet rendering as sprites in the dialogs, akin to the one in Hatsune Miku: Colorful Stage.

**Comment by fxtressia** [comment]
Also, there is a blocking issue with this code.
I'm investigating it tomorrow.

File: src/renderers/opengl/app.rs (hunk 5 above)
Comment (Speykious) [reply src/renderers/opengl/app.rs:161]: If we get a Camera struct, then I think we'll get that for free.

File: Cargo.toml (hunk 2 above)
Comment (fxtressia) [reply Cargo.toml]: All the renderers are enabled for development purposes. Please remind me to fix this before merging.

Also, it should have gone to #6 instead. Sorry!

**Comment by Speykious** [comment]
Don't investigate too deep just yet, because I've made lots of changes over at #3... Pretty much every single thing is changing.

**Review by Speykious — APPROVED** [review]
LGTM

**Comment by Speykious** [comment]
Nvm. Window doesn't close, so that would be a regression. I'll see what causes this and fix it before merging.

**Review by Speykious — CHANGES_REQUESTED** [review]
Bug: window doesn't close

**Review by Speykious — APPROVED** [review]
Window closing bug fixed, now definitely LGTM

**Pull request merged**